lib/plottr_components/src/components/project/BookList.js

```diff
@@ -57,6 +60,74 @@ const BookListConnector = (connector) => {
       beatActions.addBeat(newBookId)
     }
 
+    addBookFromTemplate = (template) => {
```

Comment:
This is very hairy and definitely breaks the abstraction.

There are two things that I think would improve the situation:

1. I think that we should add support for this in the reducers rather than the components.
I think that it would be fair (in terms of abstraction) to place the burden on the reducer to know the current structure of beats in a template and to handle adding a new book from that template.
So how about creating a new action `ADD_BOOK_FROM_TEMPLATE` which receives the template as the payload.
The reducers each have to pick off their bit of state from the template and make adjustments accordingly.
Then the components go back to having no idea that it's a beat tree and that detail is hidden away in `pltr`.

2. Clearly there's a lot of work to add a whole beat tree in terms of minting beat ids.
To solve this problem, how about finding the next id: https://github.com/cameronsutter/plottr_electron/blob/e5a65eeb6a9e571770f2ae6acdbe8244391a40f2/lib/pltr/v2/helpers/beats.js#L86 (it always produces 1 + maximum id across all beat trees) and then adding it to all the beats in the tree by reducing over the tree (https://github.com/cameronsutter/plottr_electron/blob/e5a65eeb6a9e571770f2ae6acdbe8244391a40f2/lib/pltr/v2/reducers/tree.js#L205).

Here's what that might look like (I'm writing this in the PR so please don't use it verbatim!) :)

In `reducers/tree.js` add a `parentId` parameter to the callback function in `reduce` (and add a note to state that the `reduce` function must process parents to leaves in-order.
```javascript
// Contract: reduce must process the nodes in the supplied tree, from the root node, down to the children.
// This is important to the consumers of this function who might want to process the tree, knowing, that all
// ancestors of the current node were already processed.
export const reduce = (idProp) => (tree, f, initialValue) => {
  function iter(acc, nodeId, parentId) {
    const next = nodeId === null ? acc : f(acc, findNode(tree, nodeId))
    return children(tree, nodeId).reduce((acc, entity) => {
      return iter(acc, entity[idProp], nodeId)
    }, next)
  }

  return iter(initialValue, null, null)
}
```

In `reducers/beats.js`
```javascript
case ADD_BOOK_FROM_TEMPLATE:
  const beats = action.template.beats
  const idMap = {}
  let nextId = nextId(state)
  tree.reduce('id')(beats, (newBeatTree, nextBeat, parentId) => {
    const newId = nextId
    idMap[nextBeat.id] = newId
    nextId++
    const newParentId = idMap[parentId] || null
    const newBeat = {
      ...clone(nextBeat),
      id: nextId,
    }
    return tree.addNode('id')(newBeatTree, newParentId, newBeat)
  }, tree.newTree('id'))
```

Reply:
@Quiescent so i'm trying to implement this and i think one of the wrinkles that makes me think we need to do this outside of reducers (even though that would be the best if we could) is how to use the idMap that gets created for both lines and beats in the cards reducer?

To explain that more: since we have to change the IDs of beats and lines, we also have to update those IDs in the cards that come from the template. As far as I can tell, there's no way for the cards reducer to know what those new ids are unless we do the whole thing outside reducers

Reply:
ok, never mind. After thinking it through some more, i got what you said about adding the next id to the current id and got it working. This is the comment at the root reducer:

```
// cards from the template need to know the new ids of lines and beats from the template
// the strategy here is to use the state's next id value + the template id's current value
// the card reducer will have access to the state's next id value so it will be able to determine the correct id
```

Reply:
Cool.  You're right about needing something in the root reducer to tie this all together.  :+1:

I'll take a closer look through the PR now and check to see if I can find problems with the implementation details.

Aside: there are similar problems here as we ran into when we tried to add Plot Lines from templates.  We could use some of the logic here to implement that feature too.  We still haven't resolved what it *means* to add a template based on an act structure into an existing act structure.
Do we:
 - put the two structures side by side?
 - embed the one structure into the other?
And what happens when there are too many beats?  (With the act structure off, we added a row of cards as pseudo beat titles, but that wouldn't work in an act structure).